Snailtrail 2019

li1 opened the pull request:
This PR

- updates Rust to 2018 edition
- updates most of the dependencies used, in particular timely
- reformats using cargo fmt
- fixes linting errors highlighted by cargo clippy / Rust 2018

It does _not_ touch

- the python build scripts
- tensorflow / spark adapters

since I have severely pruned / removed them on my fork for now.

antiguru commented:
Cool, thank you! I had a look and in general I think the changes make sense. There are two things I would consider:
- The new Timely API doesn't supply owned data but just a reference. To re-use allocations, use `data.swap(&mut buffer)` instead of allocating a new `Vec`, even if it's immediately drained afterwards.
- I'm not convinced the cargo fmt changes do any good. At least, they clobber the history... Do you think you can produce a version without?

li1 commented:
Thanks for your feedback!

I always thought that ˋreplaceˋ didn‘t allocate either [1], so that‘s good
to know. I‘ll update the code.

I can produce a version without ˋcargo fmtˋ changes, but I do believe that
it makes it easier to keep consistent code style throughout the project in
the long term. It can also be very non-intrusive (apart from the first time
it‘s run) — e.g., in 3DF we use it as a git hook.

I could create a separate PR for cargo fmt on SnailTrail as it was until
now. That should declobber this PR‘s history. What do you think?

[1]
https://github.com/rust-unofficial/patterns/blob/master/idioms/mem-replace.md

On Mon 1. Apr 2019 at 18:33, Moritz Hoffmann <notifications@github.com>
wrote:

> Cool, thank you! I had a look and in general I think the changes make
> sense. There are two things I would consider:
>
>    - The new Timely API doesn't supply owned data but just a reference.
>    To re-use allocations, use data.swap(&mut buffer) instead of
>    allocating a new Vec, even if it's immediately drained afterwards.
>    - I'm not convinced the cargo fmt changes do any good. At least, they
>    clobber the history... Do you think you can produce a version without?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/strymon-system/snailtrail/pull/4#issuecomment-478651652>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AJE0WVMeTAoW3WP6HZILd5huoi-INyrZks5vcjTPgaJpZM4cWDog>
> .
>

antiguru commented:
`::std::mem::replace` doesn't allocate, but the Timely contract is to get a reference to data and replace it with an empty buffer that will be passed upstream to maybe later store data again. This pattern allows us to reuse existing allocations. It also allows for allocations and deallocations to happen on the same thread, which make life simpler for the allocator.

I guess I can live with the `cargo fmt` change, but what does @utaal think? :)

Is there any specific reason to depend on Timely master instead of 0.9? I'd rather have it track a specific version than a git branch as we don't have `Cargo.lock` files in this repo.

li1 commented:
- Changed `replace` -> `swap` (I hope that's the way you were talking about?)
- timely / differential point at 0.9 instead of master
- updated authors

li1 commented:
I've now added PR #5 which simply runs `cargo fmt` on the whole project and should clear up this PR if merged prior to it.

utaal commented:
I famously don't like process for process' sake, so I don't see the value of `cargo fmt` for a project like this (small team, small codebase, for the time-being at least); however, I don't have a strong opinion for/against.

The only real problem I'd have here is that it makes it hard to see non-formatting changes - which makes it harder to do a good review. Up to @antiguru to decide whether he'd like to see e.g. a version rebased on top of #5 to see only the code changes.

vasia commented:
Hi all,
Thanks for the PR @li1!
Regarding `cargo fmt`, I think it'll be helpful, even for a project like this. It's better to start using style guidelines early and when the team is small than realizing we need it later.
That said, porting SnailTrail to current Timely and applying style changes are two separate issues and should be addressed by separate PRs.

antiguru commented:
I'm happy to accept all changes but I'm not convinced for the `cargo fmt` changes. First of all, they don't help understanding the actual changes of this PR so it makes a code review hard. Secondly, in my opinion it doesn't lead to better formatted code. There are several places where the code was more readable before formatting it.

I previously played with `cargo fmt` and each time decided that it doesn't help and that the code is more compact and readable if formatted manually.

For this reason, I'd accept the other changes (after a separate review) but not the `cargo fmt` one.

I think the best next step would be to factor out the changes that are required for Timely compatibility and create a separate PR for them. Then, we can add some of the cosmetic changes that do not alter functionality but make our code look more modern.

li1 committed:
-> Rust 2018, timely 0.9, dependency updates

li1 commented:
I've factored out all changes related to moving to Rust 2018, timely 0.9, and updating dependencies.

I'll also close the `cargo fmt` PR.

antiguru commented on logformat/rust/Cargo.toml:
```diff
@@ -1,21 +1,23 @@
+num-traits = "^0.2"
+abomonation = "0.7"
+abomonation_derive = "0.3"
+# @TODO: replace with serde
```
We'll need support for `Abomonation` as long as we pass the types through timely.

li1 replied:
I thought we might be able to replace the `json` dependency with `serde_json` at some point, since from looking at the github it didn't seem like the `json` crate is still being actively maintained and we depend on serde at dev time anyways. That's not high prio for me though.

Could `abomonation` in theory be replaced with `Serialize / Deserialize` from `serde` when using `timely` with the `bincode` feature flag? I'm not sure how big the performance hit would be, but I know we're doing something like this in 3DF. Again, not high prio, just something I noted while updating the dependencies.

antiguru commented on snailtrail/src/exploration/single_path.rs:
```diff
@@ -69,12 +66,10 @@ let graph_stream = self; //.concat(&forward_edges); //.concat(&backward_edges);
 }
 
-pub trait TraverseFrom<G: Scope, D1: Data, K: Hash + Eq + Copy + Data + Partitioning + 'static>
+pub trait TraverseFrom<G: Scope, D1: Data + abomonation::Abomonation, K: Hash + Eq + Copy + Data + Partitioning + 'static>
```
Could use `ExchangeData` here.

antiguru commented on snailtrail/src/exploration/single_path.rs:
```diff
@@ -97,7 +92,7 @@ pub trait TraverseFrom<G: Scope, D1: Data, K: Hash + Eq + Copy + Data + Partitio
 impl<TOuter: Timestamp,
      G: Scope<Timestamp = Product<TOuter, u32>>,
-     D1: Data + Debug + Hash + Eq + Send,
+     D1: Data + Debug + Hash + Eq + Send + std::marker::Sync + abomonation::Abomonation,
```
Here again.

antiguru commented on snailtrail/src/exploration/betweenness_centrality.rs:
```diff
@@ -46,9 +46,32 @@ impl ScaleReduce for f64 {
 impl<T: Data + Eq + Hash + Copy + Debug> ExtendedData for T {}
 
+/// Methods to push stream timestamps around
+pub trait PushTime<S: Scope<Timestamp = u64>, D: Data> {
```
This looks like a fine solution to the way Snailtrail computes its metric, but it may have an performance impact. This comment is really just to keep in mind that it makes a performance comparison harder between what's published in the paper and what's still to come.

li1 replied:
I will keep it in mind, thanks! Perhaps then it will make sense for the benchmark to compare current SnailTrail vs. SnailTrail using timely 0.9 vs. SnailTrail w/ future metrics to consider such effects.

li1 committed:
traversal ExchangeData
